storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -4017,122 +4014,1889 @@ dberr_t fil_space_t::garbage_collect(bool shutdown)
+
+      if (!*prev_block)
+        return err;
+
+      fil_addr_t cur_addr=
+        flst_get_next_addr((*prev_block)->page.frame
+                             + prev_addr.boffset);
+      if (cur_addr.page != descr_block.page.id().page_no() ||
+          cur_addr.boffset != xoffset)
+        return DB_CORRUPTION;
+    }
+
+    if (next_addr.page != FIL_NULL)
+    {
+      if (!node_valid(&next_addr))
+        return DB_CORRUPTION;
+
+      *next_block= fsp_get_latched_page(page_id_t{0, next_addr.page},
+                                        mtr, &err);
+      ut_ad(!*next_block == (err != DB_SUCCESS));
+      if (!*next_block)
+        return err;
+
+      fil_addr_t cur_addr=
+        flst_get_prev_addr((*next_block)->page.frame + next_addr.boffset);
+      if (cur_addr.page != descr_block.page.id().page_no() ||
+          cur_addr.boffset != xoffset)
+        return DB_CORRUPTION;
+    }
+
+    return err;
+  }
+
+  /** Complete the steps for removing the file list node
+  @param base       base block where free list starts
+  @param boffset    offset where list starts
+  @param descr      descriptor to be removed
+  @param mtr        mini-transaction */
+  static void remove_complete(buf_block_t *base, uint16_t boffset,
+                              xdes_t *descr, mtr_t *mtr) noexcept
+  {
+    fil_addr_t prev_addr= flst_get_prev_addr(descr + XDES_FLST_NODE);
+    fil_addr_t next_addr= flst_get_next_addr(descr + XDES_FLST_NODE);
+    /* remove_prepare() checked these already */
+    ut_ad(next_addr.page == FIL_NULL || node_valid(&next_addr));
+    ut_ad(prev_addr.page == FIL_NULL || node_valid(&prev_addr));
+    byte *list= base->page.frame + boffset;
+
+    buf_block_t *prev_block= nullptr;
+    buf_block_t *next_block= nullptr;
+
+    if (prev_addr.page != FIL_NULL)
+    {
+      prev_block=
+        mtr->get_already_latched(page_id_t{0, prev_addr.page},
+                                 MTR_MEMO_PAGE_SX_FIX);
+      ut_ad(prev_block);
+
+      flst_write_addr(*prev_block, prev_block->page.frame +
+                      prev_addr.boffset + FLST_NEXT,
+                      next_addr.page, next_addr.boffset, mtr);
+    }
+    else
+      flst_write_addr(*base, list + FLST_FIRST,
+                      next_addr.page, next_addr.boffset, mtr);
+
+    if (next_addr.page != FIL_NULL)
+    {
+      next_block=
+        mtr->get_already_latched(page_id_t{0, next_addr.page},
+                                 MTR_MEMO_PAGE_SX_FIX);
+      ut_ad(next_block);
+
+      flst_write_addr(*next_block, next_block->page.frame +
+                      next_addr.boffset + FLST_PREV,
+                      prev_addr.page, prev_addr.boffset, mtr);
+    }
+    else
+      flst_write_addr(*base, list + FLST_LAST,
+                      prev_addr.page, prev_addr.boffset, mtr);
+
+    /* Decrement the FLST_LEN length in LIST */
+    byte *len= list + FLST_LEN;
+    mtr->write<4>(*base, len, mach_read_from_4(len) - 1);
```

Comment:
I think that `remove_prepare()` needs to check for zero length, like `append_prepare()` is doing. The comment should say that `remove_prepare()` checked `flst_get_len()`, or similar.